Bjorncs/jvm dump

bjorncs opened the pull request:
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

bjorncs committed:
Add static factory for SyncFileInfo for service dump

bjorncs committed:
Add expiry to SyncFileInfo

bjorncs committed:
Introduce report type for service dump

bjorncs committed:
Initial implementation of Vespa service dumper

bjorncs committed:
Wire in service dumper into node agent

hakonhall commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/servicedump/ServiceDumpReport.java:
```diff
@@ -0,0 +1,68 @@
+                             @JsonProperty(EXPIRY_FIELD) Long expiry,
+                             @JsonProperty(ERROR_FIELD) String error) {
+        super(null, null);
+        this.requestedAt = requestedAt;
```
Could BaseReport's createdMillis be used instead of requestedAt?

bjorncs committed:
Export package 'com.yahoo.vespa.hosted.node.admin.maintenance.servicedump'

hakonhall commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/servicedump/VespaServiceDumperImpl.java:
```diff
@@ -0,0 +1,163 @@
+public class VespaServiceDumperImpl implements VespaServiceDumper {
+
+    private static final Logger log = Logger.getLogger(VespaServiceDumperImpl.class.getName());
+    private static final String DIRECTORY_IN_NODE = "/opt/vespa/tmp/vespa-service-dump";
```
Use NodeAgentContext::pathInNodeUnderVespaHome to avoid hard-coding Vespa home.

hakonhall commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/servicedump/VespaServiceDumperImpl.java:
```diff
@@ -0,0 +1,163 @@
+        NodeSpec nodeSpec = context.node();
+        ServiceDumpReport request = nodeSpec.reports().getReport(ServiceDumpReport.REPORT_ID, ServiceDumpReport.class)
+                .orElse(null);
+        if (request == null || request.failedAt() != null || request.completedAt() != null) {
```
Add a method on ServiceDumpReport that returns true if it has reached a terminal state?

bjorncs committed:
Rename field 'expiry' => 'expireAt'

hakonhall commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/servicedump/VespaServiceDumperImpl.java:
```diff
@@ -0,0 +1,163 @@
+            Files.deleteIfExists(directoryOnHost);
+            Files.createDirectory(directoryOnHost);
+            CommandResult result = container.executeCommandInContainerAsRoot(
+                    context, "/opt/vespa/bin/vespa-jvm-dumper", configId, DIRECTORY_IN_NODE);
```
This should use pathInNodeUnderVespaHome

hakonhall commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/servicedump/VespaServiceDumperImpl.java:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    private static String createDumpId(ServiceDumpReport report) {
+        String sanitizedConfigId = report.configId()
```
Make sure all characters of the config ID is sanitized?  This ends up as part of URI, so it sounds like it ought to be urlencoded at some point, unless the characters can be guaranteed to be valid pathname chars.

hakonhall reviewed (commented):
I'll do a full review later

bjorncs committed:
Treat value '0' as null for timestamps

bjorncs committed:
fixup! Treat value '0' as null for timestamps

bjorncs committed:
Use existing 'createdMillis' instead of 'requestedAt'

bjorncs committed:
Don't hardcode path to Vespa home

bjorncs committed:
Move check for terminal state to ServiceDumpReport

bjorncs committed:
Escape all non-alphanumeric characters

hakonhall commented on node-admin/src/test/java/com/yahoo/vespa/hosted/node/admin/maintenance/servicedump/VespaServiceDumperImplTest.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * @author bjorncs
+ */
+class VespaServiceDumperImplTest {
```
Add test of processServiceDumpRequest?

hakonhall commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/servicedump/VespaServiceDumperImpl.java:
```diff
@@ -0,0 +1,170 @@
+            CommandResult result = container.executeCommandInContainerAsRoot(
+                    context, vespaJvmDumper.toString(), configId, directoryInNode.toString());
+            context.log(log, Level.FINE, "vespa-jvm-dumper exit code: " + result.getExitCode());
+            context.log(log, Level.FINE, "vespa-jvm-dumper output: " + result.getOutput());
```
Join with above log statement?

hakonhall commented on node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/servicedump/VespaServiceDumperImpl.java:
```diff
@@ -0,0 +1,170 @@
+            URI destination = serviceDumpDestination(nodeSpec, createDumpId(request));
+            context.log(log, Level.FINE, "Uploading files with destination " + destination + " and expiry " + expiry);
+            List<SyncFileInfo> files = dumpFiles(directoryOnHost, destination, expiry);
+            logFilesToUpload(context, files);
```
We should log the files being uploaded at Level.INFO.  And it would be nice if the message would contain the actual list of files being uploaded, as sync() may choose to skip some.

bjorncs committed:
Combine log entries

bjorncs committed:
Improve logging

Log more as info. Remove redudant logging of files (already done by SyncClient).

bjorncs committed:
Don't serialize return value of getter

bjorncs committed:
Inject clock to control it in unit tests

bjorncs committed:
Make method public accessable

bjorncs committed:
Include exception name in error message + split handleFailure

bjorncs committed:
Test processServiceDumpRequest()

bjorncs commented:
@hakonhall PTAL

bjorncs committed:
Log script output as level 'info'

bjorncs committed:
Proper cleanup and permissions for temp directory

bjorncs committed:
Add trailing '/' to ensure URI paths are resolved correctly

bjorncs committed:
Compress files with '.hprof' extension

bjorncs committed:
Rename output file to 'jvm-heap-dump.hprof'

bjorncs committed:
Add '.txt' extension to text files
